File: skills/codex-review/lib/workflow-parser.ts
export type Step = {
  comments: ReadonlyArray<string>;
  env: ReadonlyArray<string>;
  uses: string | null;
  with: ReadonlyMap<string, string>;
};

export type Job = {
  comments: ReadonlyArray<string>;
  env: ReadonlyArray<string>;
  environment: string | null;
  if: string | null;
  name: string;
  permissions: ReadonlyMap<string, string>;
  steps: ReadonlyArray<Step>;
};

export type Workflow = {
  comments: ReadonlyArray<string>;
  jobs: ReadonlyArray<Job>;
  on: ReadonlyArray<string>;
  text: string;
};

const SELF_USES = /milanhorvatovic\/codex-ai-code-review-action(?:\/(prepare|review|publish))?@([0-9a-f]{40}|v\d[^\s#]*|[\w./-]+)/;

export function parseWorkflow(text: string): Workflow {
  const lines = text.split(/\r?\n/);
  const comments = collectAllComments(lines);
  const on = parseOn(lines);
  const jobs = parseJobs(lines);
  return { comments, jobs, on, text };
}

function collectAllComments(lines: ReadonlyArray<string>): string[] {
  const out: string[] = [];
  for (const line of lines) {
    const idx = indexOfComment(line);
    if (idx === -1) continue;
    out.push(line.slice(idx + 1).trim());
  }
  return out;
}

function indexOfComment(line: string): number {
  let inSingle = false;
  let inDouble = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line.charAt(i);
    if (ch === "'" && !inDouble) inSingle = !inSingle;
    else if (ch === '"' && !inSingle) inDouble = !inDouble;
    else if (ch === "#" && !inSingle && !inDouble) return i;
  }
  return -1;
}

function parseOn(lines: ReadonlyArray<string>): string[] {
  const out: string[] = [];
  let inOn = false;
  let baseIndent: number | null = null;
  for (const line of lines) {
    if (/^on:\s*$/.test(line)) {
      inOn = true;
      continue;
    }
    if (/^on:\s+\[/.test(line)) {
      const m = /^on:\s+\[(.+)\]/.exec(line);
      if (m !== null && m[1] !== undefined) {
        for (const event of m[1].split(",")) out.push(event.trim());
      }
      return out.sort();
    }
    if (!inOn) continue;
    if (line.length === 0) continue;
    const indent = line.match(/^ */)?.[0].length ?? 0;
    if (indent === 0) {
      inOn = false;
      continue;
    }
    if (baseIndent === null) baseIndent = indent;
    if (indent === baseIndent) {
      const stripped = line.slice(indent);
      const m = /^([\w-]+):/.exec(stripped);
      if (m !== null && m[1] !== undefined) out.push(m[1]);
    }
  }
  return out.sort();
}

function parseJobs(lines: ReadonlyArray<string>): Job[] {
  const jobs: Job[] = [];
  let inJobs = false;
  let baseIndent: number | null = null;
  let current: { name: string; indent: number; lines: string[] } | null = null;

  function commit(): void {
    if (current === null) return;
    jobs.push(parseJob(current.name, current.indent, current.lines));
    current = null;
  }

  for (const line of lines) {
    if (/^jobs:\s*$/.test(line)) {
      inJobs = true;
      continue;
    }
    if (!inJobs) continue;
    const trimmed = line.replace(/\s+$/, "");
    if (trimmed.length === 0) {
      if (current !== null) current.lines.push(line);
      continue;
    }
    const indent = line.match(/^ */)?.[0].length ?? 0;
    if (indent === 0 && trimmed !== "") {
      commit();
      inJobs = false;
      continue;
    }
    if (baseIndent === null) baseIndent = indent;

    if (indent === baseIndent) {
      const stripped = line.slice(indent);
      const m = /^([\w-]+):\s*$/.exec(stripped);
      if (m !== null && m[1] !== undefined) {
        commit();
        current = { indent, lines: [], name: m[1] };
        continue;
      }
    }
    if (current !== null) current.lines.push(line);
  }
  commit();
  return jobs;
}

function parseJob(name: string, jobIndent: number, lines: ReadonlyArray<string>): Job {
  const fieldIndent = jobIndent + 2;
  let ifExpr: string | null = null;
  let environment: string | null = null;
  const permissions = new Map<string, string>();
  const env: string[] = [];
  const steps: Step[] = [];
  const comments: string[] = [];

  let i = 0;
  while (i < lines.length) {
    const raw = lines[i] ?? "";
    if (raw.length === 0) {
      i++;
      continue;
    }
    const cmt = indexOfComment(raw);
    if (cmt !== -1) comments.push(raw.slice(cmt + 1).trim());
    const indent = raw.match(/^ */)?.[0].length ?? 0;
    if (indent < fieldIndent) {
      i++;
      continue;
    }
    if (indent !== fieldIndent) {
      i++;
      continue;
    }
    const stripped = raw.slice(fieldIndent).replace(/\s+#.*$/, "");
    const ifMatch = /^if:\s*(.*)$/.exec(stripped);
    if (ifMatch !== null && ifMatch[1] !== undefined) {
      ifExpr = ifMatch[1].trim();
      i++;
      continue;
    }
    const envMatch = /^environment:\s*(.*)$/.exec(stripped);
    if (envMatch !== null && envMatch[1] !== undefined && envMatch[1].trim().length > 0) {
      environment = envMatch[1].trim().replace(/^["']|["']$/g, "");
      i++;
      continue;
    }
    if (/^permissions:\s*$/.test(stripped)) {
      i++;
      while (i < lines.length) {
        const r = lines[i] ?? "";
        const ind = r.match(/^ */)?.[0].length ?? 0;
        if (ind <= fieldIndent || r.trim() === "") {
          if (ind <= fieldIndent && r.trim() !== "") break;
          if (r.trim() === "") {
            i++;
            continue;
          }
          break;
        }
        const sub = r.slice(ind).replace(/\s+#.*$/, "").trim();
        const pm = /^([\w-]+):\s*(\S+)\s*$/.exec(sub);
        if (pm !== null && pm[1] !== undefined && pm[2] !== undefined) permissions.set(pm[1], pm[2]);
        i++;
      }
      continue;
    }
    if (/^env:\s*$/.test(stripped)) {
      i++;
      while (i < lines.length) {
        const r = lines[i] ?? "";
        const ind = r.match(/^ */)?.[0].length ?? 0;
        if (ind <= fieldIndent && r.trim() !== "") break;
        if (r.trim() !== "") env.push(r.trim());
        i++;
      }
      continue;
    }
    if (/^steps:\s*$/.test(stripped)) {
      i++;
      const stepEntries = collectStepEntries(lines, i, fieldIndent);
      i = stepEntries.next;
      for (const entry of stepEntries.entries) steps.push(parseStep(entry));
      continue;
    }
    i++;
  }
  return { comments, env, environment, if: ifExpr, name, permissions, steps };
}

function collectStepEntries(
  lines: ReadonlyArray<string>,
  start: number,
  jobFieldIndent: number,
): { entries: ReadonlyArray<ReadonlyArray<string>>; next: number } {
  const entries: string[][] = [];
  let i = start;
  let current: string[] | null = null;
  while (i < lines.length) {
    const raw = lines[i] ?? "";
    if (raw.length === 0) {
      if (current !== null) current.push(raw);
      i++;
      continue;
    }
    const indent = raw.match(/^ */)?.[0].length ?? 0;
    if (indent <= jobFieldIndent && raw.trim() !== "") break;
    const stripped = raw.slice(indent);
    if (/^- /.test(stripped)) {
      if (current !== null) entries.push(current);
      current = [stripped.slice(2)];
    } else if (current !== null) {
      const stepIndent = indent;
      void stepIndent;
      current.push(stripped);
    }
    i++;
  }
  if (current !== null) entries.push(current);
  return { entries, next: i };
}

function parseStep(stepLines: ReadonlyArray<string>): Step {
  const comments: string[] = [];
  const env: string[] = [];
  let uses: string | null = null;
  const withMap = new Map<string, string>();

  let inWith = false;
  for (let i = 0; i < stepLines.length; i++) {
    const raw = stepLines[i] ?? "";
    const cmt = indexOfComment(raw);
    if (cmt !== -1) comments.push(raw.slice(cmt + 1).trim());
    const stripped = raw.replace(/\s+#.*$/, "").replace(/^\s+/, "");
    if (stripped.length === 0) continue;

    const usesMatch = /^uses:\s*(\S+)/.exec(stripped);
    if (usesMatch !== null && usesMatch[1] !== undefined) {
      uses = usesMatch[1];
      inWith = false;
      continue;
    }
    if (/^with:\s*$/.test(stripped)) {
      inWith = true;
      continue;
    }
    if (/^env:\s*$/.test(stripped)) {
      inWith = false;
      continue;
    }
    if (inWith) {
      const m = /^([\w-]+):\s*(.*)$/.exec(stripped);
      if (m !== null && m[1] !== undefined) {
        const value = stripQuotes((m[2] ?? "").trim());
        withMap.set(m[1], value);
      }
    }
    if (/\${{\s*secrets\./.test(raw) || /OPENAI_API_KEY/.test(raw)) {
      env.push(raw.trim());
    }
  }
  return { comments, env, uses, with: withMap };
}

function stripQuotes(value: string): string {
  if (value.length >= 2) {
    const first = value.charAt(0);
    const last = value.charAt(value.length - 1);
    if ((first === '"' && last === '"') || (first === "'" && last === "'")) {
      return value.slice(1, -1);
    }
  }
  return value;
}

export function selfPin(line: string): { sub: string | null; ref: string } | null {
  const m = SELF_USES.exec(line);
  if (m === null) return null;
  return { ref: m[2] ?? "", sub: m[1] ?? null };
}

export function findStepsUsingSelf(workflow: Workflow): Array<{ job: string; sub: string | null; ref: string }> {
  const out: Array<{ job: string; sub: string | null; ref: string }> = [];
  for (const job of workflow.jobs) {
    for (const step of job.steps) {
      if (step.uses === null) continue;
      const pin = selfPin(step.uses);
      if (pin === null) continue;
      out.push({ job: job.name, ref: pin.ref, sub: pin.sub });
    }
  }
  return out;
}

File: skills/codex-review/lib/workflow_parser.py
"""Workflow YAML parser tuned for the consumer-controls invariants.

The full GitHub Actions workflow grammar is large; we only parse what the CC-
predicates need: top-level on:, jobs:, per-job if/permissions/environment/env,
and per-step uses/with/env. Hand-rolled so the skill stays stdlib-only.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field


@dataclass(frozen=True)
class Step:
    comments: tuple[str, ...]
    env: tuple[str, ...]
    uses: str | None
    with_: dict[str, str]


@dataclass(frozen=True)
class Job:
    comments: tuple[str, ...]
    env: tuple[str, ...]
    environment: str | None
    if_: str | None
    name: str
    permissions: dict[str, str]
    steps: tuple[Step, ...]


@dataclass(frozen=True)
class Workflow:
    comments: tuple[str, ...]
    jobs: tuple[Job, ...]
    on: tuple[str, ...]
    text: str


_SELF_USES = re.compile(
    r"milanhorvatovic/codex-ai-code-review-action(?:/(prepare|review|publish))?@([0-9a-f]{40}|v\d[^\s#]*|[\w./-]+)"
)


def _index_of_comment(line: str) -> int:
    in_single = False
    in_double = False
    for i, ch in enumerate(line):
        if ch == "'" and not in_double:
            in_single = not in_single
        elif ch == '"' and not in_single:
            in_double = not in_double
        elif ch == "#" and not in_single and not in_double:
            return i
    return -1


def _collect_all_comments(lines: list[str]) -> tuple[str, ...]:
    out: list[str] = []
    for line in lines:
        idx = _index_of_comment(line)
        if idx == -1:
            continue
        out.append(line[idx + 1 :].strip())
    return tuple(out)


def _parse_on(lines: list[str]) -> tuple[str, ...]:
    out: list[str] = []
    in_on = False
    base_indent: int | None = None
    for line in lines:
        if re.match(r"^on:\s*$", line):
            in_on = True
            continue
        m = re.match(r"^on:\s+\[(.+)\]", line)
        if m is not None:
            for event in m.group(1).split(","):
                out.append(event.strip())
            return tuple(sorted(out))
        if not in_on:
            continue
        if line == "":
            continue
        leading = len(line) - len(line.lstrip(" "))
        if leading == 0:
            in_on = False
            continue
        if base_indent is None:
            base_indent = leading
        if leading == base_indent:
            stripped = line[leading:]
            event_match = re.match(r"^([\w-]+):", stripped)
            if event_match is not None:
                out.append(event_match.group(1))
    return tuple(sorted(out))


@dataclass
class _PendingJob:
    name: str
    indent: int
    lines: list[str] = field(default_factory=list)


def _parse_jobs(lines: list[str]) -> tuple[Job, ...]:
    jobs: list[Job] = []
    in_jobs = False
    base_indent: int | None = None
    current: _PendingJob | None = None

    def commit() -> None:
        nonlocal current
        if current is None:
            return
        jobs.append(_parse_job(current.name, current.indent, current.lines))
        current = None

    for line in lines:
        if re.match(r"^jobs:\s*$", line):
            in_jobs = True
            continue
        if not in_jobs:
            continue
        trimmed = line.rstrip()
        if trimmed == "":
            if current is not None:
                current.lines.append(line)
            continue
        leading = len(line) - len(line.lstrip(" "))
        if leading == 0 and trimmed != "":
            commit()
            in_jobs = False
            continue
        if base_indent is None:
            base_indent = leading
        if leading == base_indent:
            stripped = line[leading:]
            job_match = re.match(r"^([\w-]+):\s*$", stripped)
            if job_match is not None:
                commit()
                current = _PendingJob(name=job_match.group(1), indent=leading)
                continue
        if current is not None:
            current.lines.append(line)
    commit()
    return tuple(jobs)


def _parse_job(name: str, job_indent: int, lines: list[str]) -> Job:
    field_indent = job_indent + 2
    if_expr: str | None = None
    environment: str | None = None
    permissions: dict[str, str] = {}
    env: list[str] = []
    steps: list[Step] = []
    comments: list[str] = []

    i = 0
    n = len(lines)
    while i < n:
        raw = lines[i]
        if raw == "":
            i += 1
            continue
        cmt_idx = _index_of_comment(raw)
        if cmt_idx != -1:
            comments.append(raw[cmt_idx + 1 :].strip())
        leading = len(raw) - len(raw.lstrip(" "))
        if leading != field_indent:
            i += 1
            continue
        stripped = re.sub(r"\s+#.*$", "", raw[field_indent:])
        if_match = re.match(r"^if:\s*(.*)$", stripped)
        if if_match is not None:
            if_expr = if_match.group(1).strip()
            i += 1
            continue
        env_match = re.match(r"^environment:\s*(.*)$", stripped)
        if env_match is not None and env_match.group(1).strip():
            environment = env_match.group(1).strip().strip("\"'")
            i += 1
            continue
        if re.match(r"^permissions:\s*$", stripped):
            i += 1
            while i < n:
                row = lines[i]
                row_indent = len(row) - len(row.lstrip(" "))
                if row.strip() == "":
                    i += 1
                    continue
                if row_indent <= field_indent:
                    break
                sub = re.sub(r"\s+#.*$", "", row[row_indent:]).strip()
                pm = re.match(r"^([\w-]+):\s*(\S+)\s*$", sub)
                if pm is not None:
                    permissions[pm.group(1)] = pm.group(2)
                i += 1
            continue
        if re.match(r"^env:\s*$", stripped):
            i += 1
            while i < n:
                row = lines[i]
                row_indent = len(row) - len(row.lstrip(" "))
                if row.strip() != "" and row_indent <= field_indent:
                    break
                if row.strip() != "":
                    env.append(row.strip())
                i += 1
            continue
        if re.match(r"^steps:\s*$", stripped):
            i += 1
            entries, i = _collect_step_entries(lines, i, field_indent)
            for entry in entries:
                steps.append(_parse_step(entry))
            continue
        i += 1
    return Job(
        comments=tuple(comments),
        env=tuple(env),
        environment=environment,
        if_=if_expr,
        name=name,
        permissions=permissions,
        steps=tuple(steps),
    )


def _collect_step_entries(lines: list[str], start: int, job_field_indent: int) -> tuple[list[list[str]], int]:
    entries: list[list[str]] = []
    current: list[str] | None = None
    i = start
    n = len(lines)
    while i < n:
        raw = lines[i]
        if raw == "":
            if current is not None:
                current.append(raw)
            i += 1
            continue
        leading = len(raw) - len(raw.lstrip(" "))
        if leading <= job_field_indent and raw.strip() != "":
            break
        stripped = raw[leading:]
        if stripped.startswith("- "):
            if current is not None:
                entries.append(current)
            current = [stripped[2:]]
        elif current is not None:
            current.append(stripped)
        i += 1
    if current is not None:
        entries.append(current)
    return entries, i


def _strip_quotes(value: str) -> str:
    if len(value) >= 2 and value[0] == value[-1] and value[0] in ('"', "'"):
        return value[1:-1]
    return value


def _parse_step(step_lines: list[str]) -> Step:
    comments: list[str] = []
    env_lines: list[str] = []
    uses: str | None = None
    with_map: dict[str, str] = {}
    in_with = False

    for raw in step_lines:
        cmt_idx = _index_of_comment(raw)
        if cmt_idx != -1:
            comments.append(raw[cmt_idx + 1 :].strip())
        stripped = re.sub(r"\s+#.*$", "", raw).lstrip()
        if stripped == "":
            continue
        uses_match = re.match(r"^uses:\s*(\S+)", stripped)
        if uses_match is not None:
            uses = uses_match.group(1)
            in_with = False
            continue
        if re.match(r"^with:\s*$", stripped):
            in_with = True
            continue
        if re.match(r"^env:\s*$", stripped):
            in_with = False
            continue
        if in_with:
            kv = re.match(r"^([\w-]+):\s*(.*)$", stripped)
            if kv is not None:
                with_map[kv.group(1)] = _strip_quotes(kv.group(2).strip())
        if "${{" in raw and "secrets." in raw:
            env_lines.append(raw.strip())
        elif "OPENAI_API_KEY" in raw:
            env_lines.append(raw.strip())
    return Step(comments=tuple(comments), env=tuple(env_lines), uses=uses, with_=with_map)


def parse_workflow(text: str) -> Workflow:
    lines = text.splitlines()
    return Workflow(
        comments=_collect_all_comments(lines),
        jobs=_parse_jobs(lines),
        on=_parse_on(lines),
        text=text,
    )


@dataclass(frozen=True)
class SelfPin:
    sub: str | None
    ref: str


def self_pin(line: str) -> SelfPin | None:
    match = _SELF_USES.search(line)
    if match is None:
        return None
    return SelfPin(sub=match.group(1), ref=match.group(2) or "")


@dataclass(frozen=True)
class StepUsingSelf:
    job: str
    sub: str | None
    ref: str


def find_steps_using_self(workflow: Workflow) -> tuple[StepUsingSelf, ...]:
    out: list[StepUsingSelf] = []
    for job in workflow.jobs:
        for step in job.steps:
            if step.uses is None:
                continue
            pin = self_pin(step.uses)
            if pin is None:
                continue
            out.append(StepUsingSelf(job=job.name, sub=pin.sub, ref=pin.ref))
    return tuple(out)
